rust/agama-cli/src/auth.rs

```diff
@@ -0,0 +1,246 @@
+    Some(home::home_dir()?.join(DEFAULT_JWT_FILE))
+}
+
+/// Reads first line from given file
+fn read_line_from_file(path: &Path) -> io::Result<String> {
+    if !&path.exists() {
+        return Err(io::Error::new(
+            io::ErrorKind::Other,
+            "Cannot find the file containing the credentials.",
+        ));
+    }
+
+    if let Ok(file) = File::open(&path) {
+        // cares only of first line, take everything. No comments
+        // or something like that supported
+        let raw = BufReader::new(file).lines().next();
+
+        if let Some(line) = raw {
+            return Ok(line?);
+        }
+    }
+
+    Err(io::Error::new(
+        io::ErrorKind::Other,
+        "Failed to open the file",
+    ))
+}
+
+/// Asks user to provide a line of input. Displays a prompt.
+fn read_credential(caption: String) -> io::Result<String> {
+    let mut cred = String::new();
+
+    println!("{}: ", caption);
+
+    io::stdin().read_line(&mut cred)?;
+    if cred.pop().is_none() || cred.is_empty() {
+        return Err(io::Error::new(
+            io::ErrorKind::Other,
+            format!("Failed to read {}", caption),
+        ));
+    }
+
+    Ok(cred)
+}
+
+/// Sets the archive owner to root:root. Also sets the file permissions to read/write for the
+/// owner only.
+fn set_file_permissions(file: &Path) -> io::Result<()> {
+    let attr = fs::metadata(file)?;
+    let mut permissions = attr.permissions();
+
+    // set the file file permissions to -rw-------
+    permissions.set_mode(DEFAULT_FILE_MODE);
+    fs::set_permissions(file, permissions)?;
+
+    Ok(())
+}
+
+/// Necessary http request header for authenticate
+fn authenticate_headers() -> HeaderMap {
+    let mut headers = HeaderMap::new();
+
+    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
+
+    headers
+}
+
+/// Query web server for JWT
+async fn get_jwt(url: String, password: String) -> anyhow::Result<String> {
+    let client = reqwest::Client::new();
+    let response = client
+        .post(url)
+        .headers(authenticate_headers())
+        .body(format!("{{\"password\": \"{}\"}}", password))
+        .send()
+        .await?;
+    let body = response
+        .json::<std::collections::HashMap<String, String>>()
+        .await?;
+    let value = body.get(&"token".to_string());
+
+    if let Some(token) = value {
+        return Ok(token.clone());
+    }
+
+    Err(anyhow::anyhow!("Failed to get JWT token"))
+}
+
+/// Logs into the installation web server and stores JWT for later use.
+async fn login(password: String) -> anyhow::Result<()> {
+    // 1) ask web server for JWT
+    let res = get_jwt(DEFAULT_AUTH_URL.to_string(), password).await?;
+
+    // 2) if successful store the JWT for later use
+    if let Some(path) = jwt_file() {
+        if let Some(dir) = path.parent() {
+            fs::create_dir_all(dir)?;
+        } else {
+            return Err(anyhow::anyhow!("Cannot store the JWT token"));
```

Comment:
I would omit "JWT" (it is redundant, as "T" stands for "token"). Something more meaningful for the user could be "authentication token" or "authentication credentials".

Reply:
ok ... messages are definitely not my strong skill